### PR TITLE
Aylyth Weapons #1

### PR DESCRIPTION
**Pull request by majorsanty**

### CODE REVIEW

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java [hunk 1]
```diff
@@ -63,15 +156,18 @@ private void aylyth_injectLootDrop(CallbackInfo ci){
...
-						effect.getEffectType() == StatusEffects.INSTANT_DAMAGE ||
-						effect.getEffectType() == StatusEffects.INSTANT_HEALTH);
-			}
+        if (!this.isPlayer()) {
+            return;
 		}
+
+        LivingEntity entity = ((LivingEntity) (Object) this);
+        if(entity instanceof PlayerEntity && ((PlayerEntity) entity).getInventory().contains(ModItems.YMPE_EFFIGY_ITEM.getDefaultStack()))
+            Arrays.stream(persistentStatusEffects).filter(entity::hasStatusEffect).map(persistentStatusEffect -> false).forEach(cir::setReturnValue);
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java:171]: This is a behavior change. Using a tag for the aforementioned persistent effects and after the player and effigy checks, check the passed `effect` arg whether it is tagged or not, returning true if it is.

File: src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+@Mixin(MilkBucketItem.class)
+public class MilkBucketItemMixin {
+    @Unique
+    private static final StatusEffect[] persistentStatusEffects = { ModPotions.CRIMSON_CURSE_EFFECT };
+
+    @Inject(method = "finishUsing", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/LivingEntity;clearStatusEffects()Z"), cancellable = true)
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java:22]: Please rewrite this as a WrapOperation, collecting all of the persistent effects from the tag mentioned before and then calling the original operation (which performs the clearing and applying other wrapped operations from other mods) then reapply all of the persistent effects.

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java [hunk 3]
```diff
@@ -6,27 +6,120 @@
...
+					this.addStatusEffect(new StatusEffectInstance(ModPotions.CRIMSON_CURSE_EFFECT, 20 * 10, 0));
+				}
+
+				return value * (isPickaxe && this.getArmor() > 10f ? 1.2f : 1f);
+			}
+
+			if(usingBlight) {
+				boolean isPickaxe = entity.getMainHandStack().isOf(ModItems.BLIGHTED_PICKAXE);
+				boolean isHoe = entity.getMainHandStack().isOf(ModItems.BLIGHTED_HOE);
+				boolean isSword = entity.getMainHandStack().isOf(ModItems.VAMPIRIC_SWORD);
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java:90]: `ModItems.VAMPIRIC_SWORD` -> `ModItems.BLIGHTED_SWORD`

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java [hunk 4]
```diff
@@ -6,27 +6,120 @@
...
+
+	@Shadow public abstract boolean addStatusEffect(StatusEffectInstance effect);
+
+	@Shadow public abstract boolean isHolding(Predicate<ItemStack> predicate);
+
 	public LivingEntityMixin(EntityType<?> type, World world) {
 		super(type, world);
 	}
+	@ModifyVariable(method = "damage", at = @At("HEAD"), argsOnly = true)
+	float damage(float value, DamageSource source) {
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java:55]: I would appreciate if anything that does not affect the return value could instead be written in its own `Blighted/Vampiric...` item class using `postHit` to, for example, apply effects and such.

File: src/main/java/moriyashiine/aylyth/common/world/ModDamageSources.java [hunk 5]
```diff
@@ -32,6 +33,10 @@ public DamageSource soulRip(PlayerEntity player) {
 		return new DamageSource(damageTypeRegistry.entryOf(ModDamageTypeKeys.SOUL_RIP), player);
 	}
 
+	public DamageSource blight() {
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/common/world/ModDamageSources.java:36]: Would you mind rewriting this with a `@Nullable PlayerEntity` parameter? This way we can attribute the damage back to a given player if possible.

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java [hunk 6]
```diff
@@ -63,15 +156,18 @@ private void aylyth_injectLootDrop(CallbackInfo ci){
 		}
 	}
 
+
+	private static final StatusEffect[] persistentStatusEffects = { ModPotions.CRIMSON_CURSE_EFFECT,
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java:160]: Please make this a tag instead

**Review by dhyces — CHANGES_REQUESTED** [review]
A few changes needed. Thank you!

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java [hunk 7]
```diff
@@ -6,27 +6,120 @@
...
 	public LivingEntityMixin(EntityType<?> type, World world) {
 		super(type, world);
 	}
+	@ModifyVariable(method = "damage", at = @At("HEAD"), argsOnly = true)
+	float damage(float value, DamageSource source) {
+		if(source.getAttacker() instanceof LivingEntity entity && !source.getAttacker().getWorld().isClient) {
+			// #TODO add particles
+			double attkDMG = entity.getAttributeValue(EntityAttributes.GENERIC_ATTACK_DAMAGE);
+			boolean usingVampiric = Stream.of(ModItems.VAMPIRIC_SWORD, ModItems.VAMPIRIC_HOE, ModItems.VAMPIRIC_AXE, ModItems.VAMPIRIC_PICKAXE).anyMatch(item -> entity.getMainHandStack().isOf(item));
+			boolean usingBlight = Stream.of(ModItems.BLIGHTED_SWORD, ModItems.BLIGHTED_HOE, ModItems.BLIGHTED_AXE, ModItems.BLIGHTED_PICKAXE).anyMatch(item -> entity.getMainHandStack().isOf(item));
```
[the remark this answers is not in the document]
Comment (majorsanty) [reply src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java:60]: Aight, i'll work on these tonight

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java (hunk 1 above)
Comment (majorsanty) [reply src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java:171]: Sort of unsure on this, will push a commit with changes for it soon

File: src/main/java/moriyashiine/aylyth/common/world/ModDamageSources.java (hunk 5 above)
Comment (majorsanty) [reply src/main/java/moriyashiine/aylyth/common/world/ModDamageSources.java:36]: pushing it in a future commit

File: src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import net.minecraft.item.MilkBucketItem;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+
+import java.util.ArrayList;
+
+@Mixin(MilkBucketItem.class)
+public class MilkBucketItemMixin {
+    @Unique
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java:22]: Clean up

File: src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    @WrapOperation(method = "finishUsing", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/LivingEntity;clearStatusEffects()Z"))
+    private boolean finishUsing(LivingEntity player, Operation<Boolean> original, @Local(argsOnly = true) ItemStack stack) {
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java:25]: Cleanup local

File: src/main/java/moriyashiine/aylyth/datagen/AylythTagProviders.java [hunk 10]
```diff
@@ -159,6 +161,21 @@ protected void configure(RegistryWrapper.WrapperLookup registries) {
             getOrCreateTagBuilder(ItemTags.AXES).add(ModItems.NEPHRITE_AXE, ModItems.VAMPIRIC_AXE, ModItems.BLIGHTED_AXE);
             getOrCreateTagBuilder(ItemTags.HOES).add(ModItems.NEPHRITE_HOE, ModItems.VAMPIRIC_HOE, ModItems.BLIGHTED_HOE);
             getOrCreateTagBuilder(ItemTags.MUSIC_DISCS).add(ModItems.POMEGRANATE_CASSETTE);
+            getOrCreateTagBuilder(ModItemTags.VAMPIRIC_WEAPON).add(ModItems.VAMPIRIC_AXE, ModItems.VAMPIRIC_HOE, ModItems.VAMPIRIC_PICKAXE, ModItems.VAMPIRIC_SWORD);
+            getOrCreateTagBuilder(ModItemTags.BLIGHTED_WEAPON).add(ModItems.BLIGHTED_AXE, ModItems.BLIGHTED_AXE, ModItems.BLIGHTED_PICKAXE, ModItems.BLIGHTED_SWORD);
```
Comment (dhyces) [inline src/main/java/moriyashiine/aylyth/datagen/AylythTagProviders.java:165]: Axe is added twice

**Review by dhyces — CHANGES_REQUESTED** [review]
Just a few things to cleanup, but otherwise lgtm

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java (hunk 1 above)
Comment (dhyces) [reply src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java]: the change lgtm!

Comment (majorsanty) [reply src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java]: aight thx

File: src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java (hunk 4 above)
Comment (majorsanty) [reply src/main/java/moriyashiine/aylyth/mixin/LivingEntityMixin.java]: Tried adding a helper class and removing a considerable amount of the returns, behavior should be the same

File: src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java (hunk 9 above)
Comment (dhyces) [reply src/main/java/moriyashiine/aylyth/mixin/MilkBucketItemMixin.java:23]: i hope I remember this...

**Pull request merged**